SSA/Projects/InstCombine/ForLean.lean

```diff
@@ -572,6 +572,246 @@ theorem intMin_not_gt_zero : ¬ (intMin w >ₛ (0#w)):= by
+      simp
+      have hcases : (BitVec.toNat x < 2 ^ w') ∨ (BitVec.toNat x ≥ 2 ^ w') := by
+        apply lt_or_ge
+      cases hcases
+      case inl hle =>
+        simp [hle]
+        omega
+      case inr hgt =>
+        have hgt' : ¬ (BitVec.toNat x < 2 ^ w') := by linarith
+        simp at hgt
+        simp [hgt, hgt', bne, Nat.cast, NatCast.natCast, BEq.beq, Nat.beq]
+        omega
+
+theorem toInt_eq'' (w : Nat) (x : BitVec w): BitVec.toInt x = if x.toNat < (2 : Nat)^(w - 1) then (x.toNat : ℤ) else (x.toNat : ℤ) - 2^w := by
+  cases w <;> simp
+  split_ifs
+  unfold BitVec.toInt
+  simp
+  rename_i hh n
+  omega
+  rename_i hh n
+  unfold BitVec.toInt
+  simp at n
+  simp [n]
+  intros a
+  omega
+
+/-- If a bitvec's toInt is negative, then the toNat will be larger than half of the bitwidth. -/
+lemma large_of_toInt_lt_zero (w : Nat) (x : BitVec w) (hxToInt : BitVec.toInt x < 0) :
+    x.toNat ≥ (2 : Nat) ^ (w - 1) := by
+  rcases w with rfl | w'
+  case zero => simp at hxToInt
+  case succ =>
+    rw [toInt_eq''] at hxToInt
+    split_ifs at hxToInt
+    case pos h => linarith
+    case neg h =>
+      omega
+
+lemma toInt_lt_zero_of_large (w : Nat) (x : BitVec w) (hxLarge : x.toNat ≥ (2 : Nat) ^ (w - 1)) : BitVec.toInt x < 0
+    := by
+  rcases w with rfl | w'
+  case zero =>
+    simp [BitVec.toNat] at hxLarge
+  case succ =>
+    rw [toInt_eq'']
+    split_ifs
+    case pos h => omega
+    case neg h =>
+      norm_cast
+      rw [Int.subNatNat_eq_coe]
+      omega
+
+lemma toInt_lt_zero_iff_large (w : Nat) (x : BitVec w) : BitVec.toInt x < 0 ↔ x.toNat ≥ (2 : Nat) ^ (w - 1) := by
+  constructor
+  apply BitVec.large_of_toInt_lt_zero
+  apply BitVec.toInt_lt_zero_of_large
+
+/-- If a bitvec's toInt is negative, then the toNat will be larger than half of the bitwidth. -/
+lemma small_of_toInt_pos (w : Nat) (x : BitVec w) (hxToInt : BitVec.toInt x ≥ 0) :
+    x.toNat < (2 : Nat) ^ (w - 1) := by
+  rcases w with rfl | w'
+  case zero => simp [BitVec.width_zero_eq_zero]
+  case succ =>
+    rw [BitVec.toInt_eq''] at hxToInt
+    split_ifs at hxToInt
+    case pos h => linarith
+    case neg h =>
+      exfalso
+      simp_all
+      have hcontra : BitVec.toNat x < 2 ^ (Nat.succ w') :=
+        x.toFin.2
+      norm_cast at hxToInt
+      linarith
+
+lemma toInt_pos_of_small (w : Nat) (x : BitVec w) (hxsmall : x.toNat < (2 : Nat) ^ (w - 1)) : BitVec.toInt x ≥ 0 := by
+  rcases w with rfl | w'
+  case zero => simp
+  case succ =>
+    rw [BitVec.toInt_eq'']
+    split_ifs
+    norm_cast
+    simp
+
+lemma toInt_pos_iff_small (w : Nat) (x : BitVec w) : x.toNat < (2 : Nat) ^ (w - 1) ↔ BitVec.toInt x ≥ 0 := by
+  constructor
+  apply BitVec.toInt_pos_of_small
+  apply BitVec.small_of_toInt_pos
+
+lemma toInt_pos_iff_small' (w : Nat) (x : BitVec (w + 1)) : x.toNat < (2 : Nat) ^ w ↔ BitVec.toInt x ≥ 0 := by
+  apply BitVec.toInt_pos_iff_small
+
+lemma toInt_zero_iff (w : Nat) (x : BitVec w) : BitVec.toInt x = 0 ↔ x = 0 := by
```

Comment:
This should follow from something like `toInt_inj` and `toInt_zero`.

Reply:
It follow from `toInt_eq`.